### DISCUSSION TITLE
Memory leak in MQTTnet 4.1.3.436?

### DISCUSSION THREAD

**Discussion by mkb-mb** [dotnet/MQTTnet#1971] [Q&A] [answered]
I have written a very small application that subscribes on three topic filters.  My message handler for the IMqttClient.ApplicationMessageReceivedAsync event simply increments a counter each time it receives a message.

I have another application that publishes messages to our broker (Mosquitto). Many of those messages match one of the three topic filters. The small application typically receives hundreds of messages per second from the broker.

The problem is that the memory footprint of this application grows monotonically over time. Specifically, Gen 2 objects grow over time. Forcing a GC does not impact this behavior.

I connected a memory profiler and the objects that are accumulating are MQTTnet.Internal.AsyncSignal+AsyncSignalWaiter (these are classified as Event Handler Leaks by the memory profiler).

Is this a known defect? do I need to do something in my event handler to indicate that the message has been received?

I plan to try again with the latest version of MQTTnet.

**Comment by chkr1011** [comment]
Do you have more information regarding the AsyncSignal like where it is referenced from etc?
You do not need to do anything in order to mitigate this. So it might be a leak in the library.

**Comment by mkb-mb** [comment, marked as answer]
The observed problem (continuous memory growth) does not occur with version 4.3.3.952 of MQTTnet.
